### ISSUE TITLE
Use traits instead of structs

### ISSUE THREAD

**Issue by garro95** [Martin1887/oxigen#5]
The Genotype iter and into_iter methods should return a Type that implements the Iterator and IntoIterator traits and not the specific implementation of these traits for slices.
This would improve usability of the framework

**Comment by Martin1887** [comment]
I have been thinking about changing this in a 2.x branch.

However, it has implications in the performance if `Box`es are used to encapsulate the traits or makes more unconfortable the interface if generics are used (adding two more types to `Genotype` and `GeneticExecution`).

In the first version of the library I used this structs to simplify the interface and because all genotypes are arrays of things. What structs are you thinking about and what are their cases of use?

A possible use that I though was using `bitflags`, but I'm also thinking in implementing utilities to transform integers into Gray and vice versa.

Thanks.

**Comment by garro95** [comment]
Maybe what I want to do is out of the scope of this framework. I will clone it and try to adapt it. Then we'll see

**Comment by Martin1887** [comment]
Feel free to fork and create pull requests. If they are interesting can be merged into this project.

Anyway, a 2.x branch will be created in the following months, with more flexibility and other improvements.


Regards.

**Comment by garro95** [comment]
Do you mean there is already some work going on for this 2.x version? How can I know what are the improvements of this?

Regards.

**Comment by Martin1887** [comment]
Yes, I have been working in some ideas and implementations, but nothing has been pushed yet because I am experimenting and trying to get better results and performance.

The improvements are:
- Different replacement strategies, like more similar individuals or children replace parents.
- `from_iter` function instead of `FromIterator`requirement in `Genotype` to allow the modification of individuals changing their genes without modifying other relevant information. I am experimenting with the knapsack problem, where individuals have references to available items and knapsack' capacity for evaluation function but only the current items change.
- `distance` function in `Genotype` with default implementation to support that strategies.
- `parents_children` as argument of `SurvivalPressure` to support that strategies.
- `PopulationRefitness` trait, that allows to refitness the population in function of the other individuals. This is useful for things like niches, that are also provided.
- Multipoint and uniform cross functions.

These improvements are focused in find multiple solutions and exploring different search space areas.

Other improvements in which I have thinking (but not started) are:
- Gray encoding utilities to use binary genotypes.
- Using traits instead of structs as commented in `into_iter` and `iter` inside `Genotype`.
- Macros to create simpler encodings (vectors of basic types) without boilerplate.
- Python bindings (the loss of performance is critical for this).
- Hall of fame (the historical best individuals are randomly introduced for cross).

I have not worked on this for some weeks, but I expect releasing a new branch during summer.

**Comment by garro95** [comment]
Half of this is done by simply adding one trait bound `IntoIterator<Item=T>`

**Comment by Martin1887** [comment]
I had tried using `IntoIterator<Item = T>` as trait bound before, but that only provides `into_iter` function, and a function that does not take the ownership is also needed. I will continue thinking in the best interface.

**Comment by Martin1887** [comment]
2.0 version is more flexible than 1.x, but I have not made `Genotype` iterator types generic because it makes the interface much more cumbersome and a vector of genes is flexible enough. It would be overcome using trait alias `VecGenotype` like the current `Genotype` and a more generic `Genotype` trait with generic iterators, but they are not in Rust stable.

Maybe a future major version of Oxigen will apply this changes.
